### PR TITLE
CD for staging and production

### PR DESCRIPTION
**Pull request by sudheshna-donthineni**
- Github workflow to deploy to staging and prod lambda.
- Staging Lambda : https://ap-south-1.console.aws.amazon.com/lambda/home?region=ap-south-1#/functions/PortalBackendStaging
- Staging URL : portal-backend-staging.avantifellows.org

### CODE REVIEW

File: .github/workflows/deploy_to_prod.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# This workflow will update the code for the production environment
```
[the remark this answers is not in the document]
Comment (deepansh96) [reply .github/workflows/deploy_to_prod.yml:1]: please use the `.gitignore` file from the quiz backend repo
It includes everything

File: .github/workflows/deploy_to_staging.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+      - name: Set up SAM
+        uses: aws-actions/setup-sam@v1
+
+      - name: Configure AWS Credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: us-east-1
```
Comment (deepansh96) [inline .github/workflows/deploy_to_staging.yml:32]: ap-south-1

File: .github/workflows/deploy_to_staging.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+      - name: Build
+        run: sam build --use-container -t templates/staging.yaml
+
+      - name: Package
+        run: sam package --region ap-south-1 --template-file templates/staging.yaml --s3-bucket portal-backend-staging
```
Comment (deepansh96) [inline .github/workflows/deploy_to_staging.yml]: this step is not needed

File: app/main.py [hunk 4]
```diff
@@ -2,17 +2,20 @@
 from fastapi import FastAPI, Request
 from fastapi.responses import JSONResponse
 from fastapi_jwt_auth.exceptions import AuthJWTException
```
Comment (deepansh96) [inline app/main.py:4]: this is not working for me on local
can you please check what's the issue?

File: app/router/auth.py [hunk 5]
```diff
@@ -1,6 +1,6 @@
 from fastapi import APIRouter, Depends
```
Comment (deepansh96) [inline app/router/auth.py:1]: the file `app/router/__init_.py` is wrongly named. There should be two underscores not one

File: app/router/auth.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
...
 from fastapi_jwt_auth import AuthJWT
-from models import User
+from app.models import User
```
Comment (deepansh96) [inline app/router/auth.py:3]: this is the line that's causing the error

**Review by deepansh96 — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy_to_prod.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: us-east-1
+
+      - name: Build
+        run: sam build --use-container -t templates/prod.yaml
+
+      - name: Deploy
+        env:
+          JWT_SECRET_KEY: ${{ secrets.JWT_SECRET_KEY }}
+        run: sam deploy --stack-name PortalBackendStaging --s3-bucket portal-backend-staging --no-confirm-changeset --no-fail-on-empty-changeset --region ap-south-1 --capabilities CAPABILITY_IAM --parameter-overrides JwtSecretKey=$JWT_SECRET_KEY
```
[the remark this answers is not in the document]
Comment (sudheshna-donthineni) [reply .github/workflows/deploy_to_prod.yml:39]: Yeah, I actually haven't created anything for prod yet.

[the remark this answers is not in the document]
Comment (sudheshna-donthineni) [reply .github/workflows/deploy_to_prod.yml:39]: Let me just remove the prod flow for now. Won't be deploying to prod anytime soon.

File: .github/workflows/deploy_to_staging.yml (hunk 3 above)
Comment (sudheshna-donthineni) [reply .github/workflows/deploy_to_staging.yml]: yeah, I was actually trying something

File: app/main.py (hunk 4 above)
Comment (sudheshna-donthineni) [reply app/main.py:4]: Not getting any error on local for me.
Screenshot 2022-12-01 at 9 43 23 AM

File: .github/workflows/deploy_to_prod.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (sudheshna-donthineni) [reply .github/workflows/deploy_to_prod.yml]: Actually added prod settings.

**Comment by deepansh96** [comment]
- Even though you added `.gitignore`, `.DS_Store` is still present as it got committed earlier. You can google around to find how to remove a committed file.
- Your S3 bucket is named in prod as `prod-portal-backend` but in staging it's named as `portal-backend-staging`. We should be consistent with the naming scheme. Would suggest using the same naming scheme in all things. For example, as your stack name is `PortalBackendPord`, so the s3 bucket name should also be `portal-backend-prod`. Makes things easier  for us. Please delete the old bucket, create a new one with proper name, and re-deploy
- we need a way to run the app locally, so the fastapi_jwt_auth bug needs to be fixed :/

**Comment by sudheshna-donthineni** [comment]
> * Even though you added `.gitignore`, `.DS_Store` is still present as it got committed earlier. You can google around to find how to remove a committed file.
> * Your S3 bucket is named in prod as `prod-portal-backend` but in staging it's named as `portal-backend-staging`. We should be consistent with the naming scheme. Would suggest using the same naming scheme in all things. For example, as your stack name is `PortalBackendPord`, so the s3 bucket name should also be `portal-backend-prod`. Makes things easier  for us. Please delete the old bucket, create a new one with proper name, and re-deploy
> * we need a way to run the app locally, so the fastapi_jwt_auth bug needs to be fixed :/

So I did try to name a bucket `portal-backend-prod` but for some reason, it says bucket name already exists.

Screenshot 2022-12-02 at 1 51 06 PM

**Comment by pritamps** [comment]
> > * Even though you added `.gitignore`, `.DS_Store` is still present as it got committed earlier. You can google around to find how to remove a committed file.
> > * Your S3 bucket is named in prod as `prod-portal-backend` but in staging it's named as `portal-backend-staging`. We should be consistent with the naming scheme. Would suggest using the same naming scheme in all things. For example, as your stack name is `PortalBackendPord`, so the s3 bucket name should also be `portal-backend-prod`. Makes things easier  for us. Please delete the old bucket, create a new one with proper name, and re-deploy
> > * we need a way to run the app locally, so the fastapi_jwt_auth bug needs to be fixed :/
> 
> So I did try to name a bucket `portal-backend-prod` but for some reason, it says bucket name already exists.
> 
> Screenshot 2022-12-02 at 1 51 06 PM

This is because names need to be unique across the region, not just in AF ka AWS account. Just name is `af-portal-backend-prod` -- that should be fine.

Also, we're limited to 100 buckets total. So maybe have a bucket for `af-portal` and then `prod` and `staging` inside it as folders

**Comment by sudheshna-donthineni** [comment]
> > > * Even though you added `.gitignore`, `.DS_Store` is still present as it got committed earlier. You can google around to find how to remove a committed file.
> > > * Your S3 bucket is named in prod as `prod-portal-backend` but in staging it's named as `portal-backend-staging`. We should be consistent with the naming scheme. Would suggest using the same naming scheme in all things. For example, as your stack name is `PortalBackendPord`, so the s3 bucket name should also be `portal-backend-prod`. Makes things easier  for us. Please delete the old bucket, create a new one with proper name, and re-deploy
> > > * we need a way to run the app locally, so the fastapi_jwt_auth bug needs to be fixed :/
> > 
> > 
> > So I did try to name a bucket `portal-backend-prod` but for some reason, it says bucket name already exists.
> > Screenshot 2022-12-02 at 1 51 06 PM
> 
> This is because names need to be unique across the region, not just in AF ka AWS account. Just name is `af-portal-backend-prod` -- that should be fine.
> 
> Also, we're limited to 100 buckets total. So maybe have a bucket for `af-portal` and then `prod` and `staging` inside it as folders

Done!

**Review by Bahugunajii — APPROVED** [review]

File: .github/workflows/deploy_to_staging.yml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: ap-south-1
+
+      - name: Build
+        run: sam build --use-container -t templates/staging.yaml
+
+      - name: Deploy
+        env:
+          JWT_SECRET_KEY: ${{ secrets.JWT_SECRET_KEY }}
```
Comment (pritamps) [inline .github/workflows/deploy_to_staging.yml:39]: Same JWT key for staging and production? Maybe one of them should be different?

File: app/models.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+from typing import Union, Dict, Optional
+
+
+class User(BaseModel):
+    id: str
+    is_user_valid: bool
+    data: Union[Dict[str, str], None]
+
+
+class SessionResponse(BaseModel):
```
Comment (pritamps) [inline app/models.py:11]: Why is this file here?

**Review by pritamps — CHANGES_REQUESTED** [review]
This PR looks like it has a lot of stuff from other PRs. Can you clean it up and include only the CI/CD part?

File: .github/workflows/deploy_to_staging.yml (hunk 8 above)
Comment (sudheshna-donthineni) [reply .github/workflows/deploy_to_staging.yml:39]: Done.
